### DISCUSSION TITLE
More details on Image.linear_gradient

### DISCUSSION THREAD

**Discussion by sacrophyte** [python-pillow/Pillow#6276] [Q&A] [answered]
Since Andrew so kindly answered my previous question about a progress bar, I am trying to learn more about linear_gradient. The docs seem a bit bland and bare - am I looking in the right place?
https://pillow.readthedocs.io/en/latest/reference/Image.html?highlight=linear_gradient#PIL.Image.linear_gradient

What is "mode"? What are the range of values?
How do you change the colors of the gradient (eg, start and stop colors)?

Or maybe this is common knowledge and I simply missing it.

**Comment by radarhere** [comment, marked as answer]
For information about mode, see https://pillow.readthedocs.io/en/latest/handbook/concepts.html#modes

`linear_gradient` is not a very flexible function. When the docs say "Generate 256x256 linear gradient from black to white, top to bottom.", that's not just the default functionality, that's the entire functionality.

To generate a horizontal gradient from red to yellow in my [previous example](https://github.com/python-pillow/Pillow/discussions/6271#discussioncomment-2693092),
- I generated the 256x256 gradient
- Cropped it to be 50 pixels wide
- Rotated it by 90 degrees, so that the gradient was horizontal instead of vertical
- Used `merge()` to have a solid red channel, treat the gradient as the green channel, and an empty blue channel. That meant that the gradient went from full red and no green to full red and full green, a.k.a. yellow.

That was just me putting together a quick example. If you want to create a more complex linear gradient, it would probably be simpler to use `putpixel()`.

```python
from PIL import Image

def gradient(size, start, stop):
    im = Image.new("RGB", size)
    for x in range(im.width):
        progress = x / (im.width - 1)
        color = tuple(int(start[i] * (1 - progress) + stop[i] * progress) for i in range(3))
        for y in range(im.height):
            im.putpixel((x, y), color)
    return im

im = gradient((256, 50), (255, 0, 0), (255, 255, 0))
im.show()
```

**Reply by sacrophyte** [reply]
Thanks, this is very helpful!

**Comment by orthello** [comment]
I used this snippet with some alterations.
I changed it to make it top to bottom (change x to y and width to height (and vice versa)
Only `im.putpixel((x, y), color)` should remain the same.

If you also want to take alpha (transparency) into account, make it an RGB**A** Image and change `range(3)` to `range(4)`
(And use RGBA color for start and stop)

```python
from PIL import Image

def gradient(size, start, stop):
    im = Image.new("RGBA", size)
    for y in range(im.height):
        progress = y / (im.height - 1)
        color = tuple(int(start[i] * (1 - progress) + stop[i] * progress) for i in range(4))
        for x in range(im.width):
            im.putpixel((x, y), color)
    return im

im = gradient((50, 256), (255, 0, 0, 20), (255, 255, 0, 255))
im.show()
```
